### PR TITLE
wizard/oscap: update the oscap step to latest specifications

### PR DESCRIPTION
**Pull request by mgold1234**
Before, in the Oscap step, the drop down select menu was only listing profile_ids
Now, it shows the profiles names. When the user selects one, the details about the
profile are displayed under the selection, improving user experience.

FIXES # - HMS-3087
Screenshot 2023-11-21 at 15 40 23

### CODE REVIEW

**Review by regexowl — APPROVED** [review]
Looks good! One teeny tiny suggestion as I'd say that the Documentation button looks better on the same line as the text, but I'll leave the final decision to you :)

File: src/Components/CreateImageWizard/steps/oscap.js
Comment (regexowl) [inline src/Components/CreateImageWizard/steps/oscap.js:31]: The documentation buttons needs to be on the same line as the text if we want to keep the space between the button and the text:

Same line:
image

Separate line: 
image


```suggestion
          Use OpenSCAP to monitor the adherence of your registered RHEL systems
          to a selected regulatory compliance profile. <DocumentationButton />
```

File: src/Components/CreateImageWizard/CreateImageWizard.scss
Comment (regexowl) [inline src/Components/CreateImageWizard/CreateImageWizard.scss:10]: This won't be needed when the description list is implemented.

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (regexowl) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:143]: The TextLists work on a principal of term / description pair. So you'll have one `TextListItemVariants.dt` for the term and `TextListItemVariants.dd` for the description. Docs for this component are here: https://www.patternfly.org/components/text/#description-list

```suggestion
            component={TextListItemVariants.dd}
```

The `className="normal-text"` can also be removed as the `dd` formatting will keep the text not bold. :)

**Review by regexowl — COMMENTED** [review]
Added a few more comments

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lucasgarfield) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:98]: We should always avoid adding css whenever possible... which is actually most of the time, fortunately. :smile: 

```
<Button variant="plain" aria-label="About OpenSCAP">
    <HelpIcon />
</Button>
```

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lucasgarfield) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:83]: Let's actually just delete this popover for now, since we don't have good copy for it.

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lucasgarfield) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:82]: This should actually be passed into `<FormGroup>` as a `label` prop, like so: `<FormGroup label={'OpenSCAP profile'} ... />`.

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lucasgarfield) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:123]: I think we should only show this `<TextContent>` if a profile has been selected.

**Comment by lucasgarfield** [comment]
We've merged the back-end changes and I have created a PR to update the API... but it isn't merged yet, and probably won't be until Monday... however you can work on this on Sunday by rebasing onto #1471. Then you should be able to get the data you need to populate the Description and Profile ID fields!

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (mgold1234) [reply src/Components/CreateImageWizard/formComponents/Oscap.js:83]: I asked Kelsea about the content of OpenScap Profile when we go over it and she gave me this content    'To run a manual compliance scan in OpenSCAP, download this image.' i will change it

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (mgold1234) [reply src/Components/CreateImageWizard/formComponents/Oscap.js]: thank you @lavocatt , I tried to add the spinner, but its seems that we need different solution here.
I added the onMouseEnter, thank you for mention that

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (mgold1234) [reply src/Components/CreateImageWizard/formComponents/Oscap.js:89]: if I added the label, the question mark moved under the label

**Comment by lavocatt** [comment]
Thanks for your changes, with the code linted and the unit tests passing it's way better to dig into your contribution. I'm now running the PR on my machine with the `msw` and I'm noticing several issues:

* there is no data showing up in the dropdown menu
* the label `initial Profile Name` won't clear up upon clicking on the cross
* I would expect that the label `initial Profile Name` would disappear when I'm clicking on the text input to type and filter what I want to see
* I would expect that the label `initial Profile Name` would be in a different color than solid black.

image

To help you figure out how to change the behavior of the component you can take inspiration on some other type ahead select that have been implemented in the wizard. For instance:

https://github.com/RedHatInsights/image-builder-frontend/blob/fabd3e015a220dbfdb7aef0b30a43f9faa5bc9c8/src/Components/CreateImageWizard/formComponents/AWSSourcesSelect.js#L91-L154

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lavocatt) [reply src/Components/CreateImageWizard/formComponents/Oscap.js]: I invite to you have a look on how it's done for other type ahead selects in the code base, for instance:

https://github.com/RedHatInsights/image-builder-frontend/blob/fabd3e015a220dbfdb7aef0b30a43f9faa5bc9c8/src/Components/CreateImageWizard/formComponents/AWSSourcesSelect.js#L91-L154

It's very important to check for loading and error states as it provides visual clues to the user that the application isn't stuck and is actually doing something.

File: src/Components/CreateImageWizard/CreateImageWizard.js
Comment (regexowl) [inline src/Components/CreateImageWizard/CreateImageWizard.js:521]: Correct me if I'm wrong, but this is not really forming a state, right?

Shouldn't there be an assignment to the state?
```suggestion
      formState['oscap-profile-name'] = composeRequest?.customizations?.openscap?.profile_name;
```

File: src/Components/CreateImageWizard/CreateImageWizard.js
Comment (regexowl) [inline src/Components/CreateImageWizard/CreateImageWizard.js:522]: Same here, I'd expect something like:
```suggestion
      formState['oscap-profile-description'] = composeRequest?.customizations?.openscap?.profile_description;
```

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (regexowl) [reply src/Components/CreateImageWizard/formComponents/Oscap.js:87]: In this case you'll need to make the popover button a part of the label, then it should be both on the same line and be styled correctly with the little star for a required field.

So something like this:
```javascript
    <FormGroup
      isRequired={true}
      data-testid="profiles-form-group"
      label={
        <>
          OpenSCAP profile
          <Popover
            maxWidth="30rem"
            position="left"
            bodyContent={
              <TextContent>
                <Text>
                  To run a manual compliance scan in OpenSCAP, download this
                  image.
                </Text>
              </TextContent>
            }
          >
            <Button variant="plain" aria-label="About OpenSCAP" isInline>
              <HelpIcon />
            </Button>
          </Popover>
        </>
      }
    >
```

will result in this:
image

You could also add `className="pf-u-p-0"` to the button to make it a bit more compact, but as we have similar spacing in few other places in the Wizard I wouldn't be bothered with it in this PR. :)

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (regexowl) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:103]: Should the select work as a typeahead select? Typing into it doesn't seem to filter the results. Also when you type something and hit enter to possible trigger the filtering the whole wizard jumps back to the first step.

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (regexowl) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:107]: When a profile is selected the select contains a reference ID instead of the profile name.

**Review by regexowl — COMMENTED** [review]
Thank you for all the work, the step is looking really neat! :sparkles: 

I've added some comments. Also wanted to ask how we are thinking about the step now. I'm not really up to date with this, but I've noticed "None" is no longer there as an profile option. Instead we have the "Select a profile" placeholder which works for me, but then we maybe don't want to mark the field as required?

The thing is, with it being required there's a situation when a profile is not selected, but the Next button is not disabled which goes against the mandatory field I think.

File: src/Components/CreateImageWizard/CreateImageWizard.js
Comment (lavocatt) [inline src/Components/CreateImageWizard/CreateImageWizard.js:522]: In this function we are loading a `composeRequest` into a state for the wizard. However these two values are not going to endup in the request, so they will never get to be loaded. It is thus safe to remove these lines.

But the fact that the data can't be loaded from a request will have side effects for the wizard.

It means that, upon recreating a request, when entering the `oscap` step, we will have the issue where the `profileName` is initialized to an empty string. 

Let's fix this :slightly_smiling_face:

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lavocatt) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:57]: In order to fix the issue, let's not use a form state here (sorry for pointing you in the wrong direction earlier). We will work with the  `useGetOscapCustomizationsQuery` hook and a `useEffect`.

The main idea is:
1. the `profileName` gets initialized with an empty string all the time.
2. `useGetOscapCustomizationsQuery` is called with the selected `profile` and distribution (same as you did at line 185)
3. then we will use a `useEffect` to propagate the data to the `profileName`

I've tested this solution on my local copy and it works well. Clicking back and forth in the wizard keeps the value, the clear button clear the field and recreating an image works fine.

If you need any help with this I'll be back on Monday to give you a hand.

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lavocatt) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:81]: We will also need to remove this call to `change`

File: src/Components/CreateImageWizard/formComponents/Oscap.js
Comment (lavocatt) [inline src/Components/CreateImageWizard/formComponents/Oscap.js:195]: And this other one too.

**Review by lavocatt — COMMENTED** [review]
Thanks for the last changes! I realized that, that during our debugging session I misguided you into using a form state, only to realize later by looking a your PR that it wouldn't work for image recreation. (you'll find more details about why in the comments below)

So we need to change the approach and use the rtk query + a useEffect to make this work correctly.

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:140]: The tests don't check the `/tmp` mount point. Which they probably should in this place. 

The structure of the test could be following:
1. get to the OpenSCAP step
2. select a profile
3. click Next
4. check if there is a mount point which is defined in the profile
5. click Next
6. check if there is/are package/s added based on the profile

File: src/test/fixtures/composes.ts
Comment (regexowl) [inline src/test/fixtures/composes.ts:323]: I'm wondering whether the mocks should use real data or data that re obviously mocked. I'm mentioning the same thing in a comment for `oscap.ts` fixture. Will start a thread about it.

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:144]: The comment does not reflect what happens in the tests.

**Review by lavocatt — APPROVED** [review]
Everything works very well. Great Job Michal!

**Comment by lavocatt** [comment]
@jrusz I think this PR is breaking the IQE tests by changing how the OpenScap step is working. Can you check it out please?

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:162]: As most of the testing runs asynchronously the first option should be in my opinion be using the combination of `await` + `findBy` query.

The difference between `getBy` and `findBy` is that `getBy` tries to get an element instantaneously and fails if the element haven't been rendered yet. This could cause a certain level of flakiness in this case.

The combination of `await` and `findBy` is able to wait for the element to be rendered and therefore doesn't throw an error in case that didn't happen immediately.

Why did the two get switched here? Was the async `findBy` causing any problems in the tests?

File: src/store/imageBuilderApi.ts
Comment (regexowl) [inline src/store/imageBuilderApi.ts:299]: This looks like a manual change to the generated code which means that when the API schema gets updated and the `npm run api` is run the change will get overwritten and lost.

If the change is needed it should be done directly in the image-builder API so it gets propagated consistently. Do the name and description need to be required?

File: src/Components/CreateImageWizard/formComponents/ReviewStepTextLists.js
Comment (regexowl) [inline src/Components/CreateImageWizard/formComponents/ReviewStepTextLists.js:632]: For consistency:

```suggestion
          Profile name:
```

**Comment by jrusz** [comment]
> @jrusz I think this PR is breaking the IQE tests by changing how the OpenScap step is working. Can you check it out please?

We don't build OpenScap images in the tests though so it should be fine. I'll check out the `pr_check` once it finishes if there is anything else broken.

**Review by regexowl — CHANGES_REQUESTED** [review]
This looks amazing. Thanks for all the work you've done on this! Still have some code to go through, but wanted to submit a first batch of comments.

**Comment by lavocatt** [comment]
> This looks amazing. Thanks for all the work you've done on this! Still have some code to go through, but wanted to submit a first batch of comments.

Good idea! Thanks for preventing a too enthusiastic merge.

File: src/store/imageBuilderApi.ts
Comment (lavocatt) [reply src/store/imageBuilderApi.ts:299]: Is it a manual change? Travis should catch it then! But I think Travis is broken :sob:

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (lavocatt) [reply src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:162]: Oh yes! Good catch! This is super important to do.

File: src/Components/CreateImageWizard/formComponents/OscapProfileInformation.tsx
Comment (regexowl) [inline src/Components/CreateImageWizard/formComponents/OscapProfileInformation.tsx:25]: The `isErrorOscapProfileInfo` seems to be unused so it can be removed.

File: src/store/imageBuilderApi.ts
Comment (regexowl) [reply src/store/imageBuilderApi.ts:299]: Great point! As we are planning to no longer depend on Travis anyway, we should probably add the check to our dev check actions. I'll make a note and take a look at it.

Comment (regexowl) [reply src/store/imageBuilderApi.ts:299]: Actually - general question about the name and description in the API - do we even use it in the code? What I mean is, I've tried to add an OSCAP profile to a mock to see what I get on Recreate.

Updated a compose fixture with:

image

But what I see on Recreate is:

image

And now I'm a tad bit confused :sweat_smile:

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (mgold1234) [reply src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:162]: if I dont change it, the test fail    await screen.findByText('button', {
          |                  ^
      168 |       name: /review/i,
      169 |     });

File: src/store/imageBuilderApi.ts
Comment (mgold1234) [reply src/store/imageBuilderApi.ts:299]: I run the npm run api and its adding those rows automatically, I didnt add it manually .

Comment (regexowl) [reply src/store/imageBuilderApi.ts:299]: Yup, the rows should be added automatically. I was wondering about the removed question marks in one of the previous commits. But that could have also been a mismatch between schema and API version before :+1:  :)

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [reply src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:162]: Hmm, it does seem to pass for me with this:

```suggestion
    await screen.findByRole('button', {
      name: /review/i,
    });
```

Wonder what the difference is. But also this shouldn't be a blocker I'd say. There should be some check and update of the tests anyway in the not so distant future as we are still using `getBy` in few other places.

**Comment by jrusz** [comment]
The whole step is not visible in ephemeral environment. Is it hidden behind a feature flag?

**Comment by regexowl** [comment]
> The whole step is not visible in ephemeral environment. Is it hidden behind a feature flag?

Yes, it should be gated behind `image-builder.wizard.oscap.enabled` (https://github.com/RedHatInsights/image-builder-frontend/pull/1392). Enabled on stage, disabled on prod I think?

File: src/store/imageBuilderApi.ts
Comment (lavocatt) [reply src/store/imageBuilderApi.ts]: Regarding the check that's done in travis, you can move it to to a github action easily, it's just a bash oneliner.

Comment (regexowl) [reply src/store/imageBuilderApi.ts]: Tadaa :tada: https://github.com/RedHatInsights/image-builder-frontend/pull/1510

**Comment by jrusz** [comment]
Ok I've updated the IQE plugin with the new step name. We don't do do anything else on that step besides clicking next currently..

**Comment by mgold1234** [comment]
thank you @jrusz

File: src/test/fixtures/composes.ts
Comment (regexowl) [inline src/test/fixtures/composes.ts:323]: Can we make this a mocked description as well?

```suggestion
            'This is a mocked profile description. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean posuere velit enim, tincidunt porttitor nisl elementum eu.',
```

File: src/test/fixtures/composes.ts
Comment (regexowl) [inline src/test/fixtures/composes.ts:824]: And this also

```suggestion
              'This is a mocked profile description. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean posuere velit enim, tincidunt porttitor nisl elementum eu.',
```

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:102]: I see, in case there would be more partitions the test would fail as there would be multiple cells with "xfs" as a content. I don't really mind it, but it might not be immediately obvious. What you can do is simply query for the tmp mount point and check it's not there when everything renders.

Which can be done like this:

```suggestion
    expect(
      screen.queryByRole('cell', {
        name: /tmp/i,
      })
    ).not.toBeInTheDocument();
```

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:130]:
```suggestion
    // Now we should be at the OpenSCAP step
```

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (regexowl) [inline src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:104]:
```suggestion
    // check that there are no Packages contained when selecting the "None" profile option
```

**Review by regexowl — CHANGES_REQUESTED** [review]
This looks great! Thank you so much :heart: 

I've added a few nitpicks, we can merge after those are in!

**Comment by regexowl** [comment]
Considering the Travis failure - from what I was able to find it could be a problem with the npm cache as the `npm ci` command seems to be failing only for Travis. Correct me if I'm wrong, but locally it's fine, right? Also there are no changes to `packages.json` so not really sure what glitched on the clean install.

I can see two solutions - we can somehow try to completely re-trigger the run inside Travis. Or since we plan to drop Travis anyway and the same check seems to be passing in dev checks we could maybe just ignore it? What do you think @lucasgarfield? Would you be okay with merging with failed Travis or should we look into it more?

**Comment by regexowl** [comment]
Although - thinking about the failing Travis CI - couldn't we just add flushing the cache and see what it does?

I mean adding a line to clean the npm cache to this PR, checking if it fixes Travis and if it does we can remove the line before merging the PR. This shouldn't affect anything outside then. Does it make sense? Not sure if I'm overlooking something.

File: package.json
Comment (regexowl) [inline package.json:65]: This looks strange. Is this an intended change @mgold1234?

File: src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js
Comment (mgold1234) [reply src/test/Components/CreateImageWizard/CreateImageWizard.compliance.test.js:146]: I covered that test in row 149   // check that the FSC contains a /tmp partition
    await clickNext();
    await screen.findByRole('heading', { name: /File system configuration/i });
    await screen.findByText(/tmp/i);

**Review by regexowl — APPROVED** [review]
This looks great! Thank you for all your work on this!! :heart:

**Comment by regexowl** [comment]
As the CI was passing when the ephemeral cluster was up and the changes since then were minimal I think it should be safe to merge this. Is that okay @jrusz?

**Comment by jrusz** [comment]
> As the CI was passing when the ephemeral cluster was up and the changes since then were minimal I think it should be safe to merge this. Is that okay @jrusz?

Yeah, I can trigger a manual test run in stage afterwards to make sure.

**Pull request merged**
